docs/howto/migrate/migration-guide.md

```diff
@@ -0,0 +1,1542 @@
+
+```scala mdoc:silent:nest
+val app: ZIO[zio.ZEnv, Nothing, Unit] = 
+  myApp.provideCustomLayer(
+    ((Logging.live ++ Database.live ++ (Logging.live >>> BlobStorage.live)) >>> DocRepo.live) ++
+      ((Logging.live ++ Database.live) >>> UserRepo.live)
+  )
+```
+
+In ZIO 2.x, the `ZIO#injectCustom` does the similar but for the injection:
+
+```scala mdoc:silent:nest
+val app: ZIO[zio.ZEnv, Nothing, Unit] =
+  myApp.injectCustom(
+    Logging.live,
+    DocRepo.live,
+    Database.live,
+    BlobStorage.live,
+    UserRepo.live
+  )
+```
+
+> Note:
+> All `provide*` methods are not deprecated, and they are still necessary and useful for low-level and custom cases. But, in ZIO 2.x, in most cases, it's easier to use `inject`/`wire` methods.
+
+
+| ZIO 1.x and 2.x (manually)                    | ZIO 2.x (automatically) |
+|-----------------------------------------------|-------------------------|
+| `ZIO#provide`                                 | `ZIO#inject`            |
+| `ZIO#provideSomeLayer`                        | `ZIO#injectSome`        |
+| `ZIO#provideCustomLayer`                      | `ZIO#injectCustom`      |
+| Composing manually using `ZLayer` combinators | `ZLayer#wire`           |
+| Composing manually using `ZLayer` combinators | `ZLayer#wireSome`       |
+
+### ZLayer Debugging
+
+To debug ZLayer construction, we have two built-in layers, i.e., `ZLayer.Debug.tree` and `ZLayer.Debug.mermaid`. For example, by including `ZLayer.Debug.mermaid` into our layer construction, the compiler generates the following debug information:
+
+```scala
+val layer = ZLayer.wire[Has[DocRepo] with Has[UserRepo]](
+  Console.live,
+  Logging.live,
+  DocRepo.live,
+  Database.live,
+  BlobStorage.live,
+  UserRepo.live,
+  ZLayer.Debug.mermaid
+)
+```
+
+```scala
+[info]   ZLayer Wiring Graph  
+[info] 
+[info] ◉ DocRepo.live
+[info] ├─◑ Logging.live
+[info] │ ╰─◑ Console.live
+[info] ├─◑ Database.live
+[info] ╰─◑ BlobStorage.live
+[info]   ╰─◑ Logging.live
+[info]     ╰─◑ Console.live
+[info] 
+[info] ◉ UserRepo.live
+[info] ├─◑ Logging.live
+[info] │ ╰─◑ Console.live
+[info] ╰─◑ Database.live
+[info] 
+[info] Mermaid Live Editor Link
+[info] https://mermaid-js.github.io/mermaid-live-editor/edit/#eyJjb2RlIjoiZ3JhcGhcbiAgICBDb25zb2xlLmxpdmVcbiAgICBCbG9iU3RvcmFnZS5saXZlIC0tPiBMb2dnaW5nLmxpdmVcbiAgICBMb2dnaW5nLmxpdmUgLS0+IENvbnNvbGUubGl2ZVxuICAgIFVzZXJSZXBvLmxpdmUgLS0+IExvZ2dpbmcubGl2ZVxuICAgIFVzZXJSZXBvLmxpdmUgLS0+IERhdGFiYXNlLmxpdmVcbiAgICBEb2NSZXBvLmxpdmUgLS0+IERhdGFiYXNlLmxpdmVcbiAgICBEb2NSZXBvLmxpdmUgLS0+IEJsb2JTdG9yYWdlLmxpdmVcbiAgICBEYXRhYmFzZS5saXZlXG4gICAgIiwibWVybWFpZCI6ICJ7XG4gIFwidGhlbWVcIjogXCJkZWZhdWx0XCJcbn0iLCAidXBkYXRlRWRpdG9yIjogdHJ1ZSwgImF1dG9TeW5jIjogdHJ1ZSwgInVwZGF0ZURpYWdyYW0iOiB0cnVlfQ==
+```
+
+### Module Pattern
+
+_Module Patternـ is one of the most significant changes in ZIO 2.x. Before discussing changes, let see how did we define services in ZIO 1.x.
+
+Let's try to write a Logging service using _Module Pattern 2.0_:
```

Comment:
> Let's try to write a Logging service using _Module Pattern 2.0_:

one might think that the code that immediately follows is from 2.0, but it looks 1.0 related

Reply:
Thanks! I've fixed that.